### PR TITLE
Refactor: Send heartbeat with dedicated workers

### PR DESCRIPTION
**Pull request by drmingdrmer**
##### Refactor: Send heartbeat with dedicated workers

Heavy AppendEntries traffic can block heartbeat messages. For example,
future AppendEntries in stream RPC may not receive a response indicating
a follower is alive. In such cases, the leader might time out to extend
its lease, and be considered partitioned from the cluster.

This commit moves heartbeat broadcasting to separate tasks that won't be
blocked by AppendEntries. This ensures the leader can always be
acknowledged with the liveness of followers.

Separate log progress notification and clock progress notification:
When ReplicationCore successfully finished one RPC to Follower/Learner,
it informs the RaftCore to update log progress and clock(heartbeat) progress.
This commit split these two informations into two `Notification`
variants, in order to make progress handling more clear.

Another improvement is to ignore a heartbeat progress if it is sent with
an older cluster membership config. Because a follower can be removed
and re-added, the obsolete heartbeat progress is invalid. This check is
done by remembering the membership log id in the `HeartbeatEvent`.

`HigherVote` can be sent directly to Notification channel.
replication::Response does not need `HigherVote` variant any more.
And `Response` is renamed to `Progress`

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/openraft/1225)

### CODE REVIEW

**Comment by drmingdrmer** [comment]
@ariesdevil

File: openraft/src/core/heartbeat/worker.rs [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+use crate::Config;
+use crate::RaftTypeConfig;
+
+/// A dedicate worker sending heartbeat to a specific follower.
+pub struct HeartbeatWorker<C, N>
+where
+    C: RaftTypeConfig,
+    N: RaftNetworkV2<C>,
+{
+    pub(crate) id: C::NodeId,
```
Comment (ariesdevil) [inline openraft/src/core/heartbeat/worker.rs:28]: Is `id` needed here?

Comment (drmingdrmer) [reply openraft/src/core/heartbeat/worker.rs:28]: For logging purpose only :(

**Review by ariesdevil — APPROVED** [review]

**Review by schreter — DISMISSED** [review]
From what I can say, looks good. Just see the comment - not sure about semantics change.

> Reviewed 5 of 16 files at r1, 13 of 13 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/datafuselabs/openraft/1225#-:-O3XNBoU5D-yw0Syor3H:bwq89o6)* status: all files reviewed, 2 unresolved discussions (waiting on @ariesdevil and @drmingdrmer)

___
*[`openraft/src/core/heartbeat/worker.rs` line 87 at r3](https://reviewable.io/reviews/datafuselabs/openraft/1225#-O3XLXXN0Z949VfMIM0w:-O3XLXXN0Z949VfMIM0x:bfvee6p) ([raw file](https://github.com/datafuselabs/openraft/blob/2411bd7a0a7a4df402dc777bdec2cd4cb47a900e/openraft/src/core/heartbeat/worker.rs#L87)):*
> ```XML
>             let payload = AppendEntriesRequest {
>                 vote: *heartbeat.session_id.leader_vote.deref(),
>                 prev_log_id: None,
> ```

Since it's now sending empty `AppendEntriesRequest` with `prev_log_id == None`, this is a bit of a semantics change.

Maybe document on append entries somewhere that the `prev_log_id` is uninteresting when the entries are empty?

I believe in our implementation we are checking `prev_log_id` (currently unconditionally).
___
*[`openraft/src/engine/handler/leader_handler/send_heartbeat_test.rs` line 60 at r3](https://reviewable.io/reviews/datafuselabs/openraft/1225#-O3XM20fCXMMso7dezOv:-O3XM20fCXMMso7dezOw:b-jncb9v) ([raw file](https://github.com/datafuselabs/openraft/blob/2411bd7a0a7a4df402dc777bdec2cd4cb47a900e/openraft/src/engine/handler/leader_handler/send_heartbeat_test.rs#L60)):*
> ```XML
>         assert_eq!(
>             vec![
>                 //
> ```

Why those empty comments (also below)?

**Review by drmingdrmer — COMMENTED** [review]
> Reviewed 1 of 16 files at r1.
> *[Reviewable](https://reviewable.io/reviews/datafuselabs/openraft/1225)* status: all files reviewed, 2 unresolved discussions (waiting on @ariesdevil and @schreter)

___
*[`openraft/src/core/heartbeat/worker.rs` line 87 at r3](https://reviewable.io/reviews/datafuselabs/openraft/1225#-O3XLXXN0Z949VfMIM0w:-O3_Qd-D3sQilRoYUrNC:b-psifkt) ([raw file](https://github.com/datafuselabs/openraft/blob/2411bd7a0a7a4df402dc777bdec2cd4cb47a900e/openraft/src/core/heartbeat/worker.rs#L87)):*
<details><summary><i>Previously, schreter wrote…</i></summary><blockquote>

Since it's now sending empty `AppendEntriesRequest` with `prev_log_id == None`, this is a bit of a semantics change.

Maybe document on append entries somewhere that the `prev_log_id` is uninteresting when the entries are empty?

I believe in our implementation we are checking `prev_log_id` (currently unconditionally).
</blockquote></details>

Semantically, `prev_log_id` is used to ensure that the log entries in `AppendEntries.entries` are consecutive with the entries on the leader by asserting that `prev_log_id` is equal. Raft ensures that a matching log ID implies all preceding logs match. If `entries` is empty, it is considered consecutive at any index on the follower, and `prev_log_id=None` matches the very first position `None`.

In short AppendEntries with `prev_log_id=None && entries.is_empty()` means to append nothing at the very beginning, which is always valid.

[https://github.com/datafuselabs/openraft/blob/ada27b447a2a89ca97a6ed5061e3cc724883de85/openraft/src/engine/handler/following_handler/mod.rs#L116](https://github.com/datafuselabs/openraft/blob/ada27b447a2a89ca97a6ed5061e3cc724883de85/openraft/src/engine/handler/following_handler/mod.rs#L114-L131)
___
*[`openraft/src/engine/handler/leader_handler/send_heartbeat_test.rs` line 60 at r3](https://reviewable.io/reviews/datafuselabs/openraft/1225#-O3XM20fCXMMso7dezOv:-O3_QQoeAbkCyuSLGI7p:b6ibmwn) ([raw file](https://github.com/datafuselabs/openraft/blob/2411bd7a0a7a4df402dc777bdec2cd4cb47a900e/openraft/src/engine/handler/leader_handler/send_heartbeat_test.rs#L60)):*
<details><summary><i>Previously, schreter wrote…</i></summary><blockquote>

Why those empty comments (also below)?
</blockquote></details>

To force line break and minimize git diff changes :(

**Review by schreter — APPROVED** [review]
> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/datafuselabs/openraft/1225)* status: all files reviewed, 1 unresolved discussion (waiting on @ariesdevil)

___
*[`openraft/src/core/heartbeat/worker.rs` line 87 at r3](https://reviewable.io/reviews/datafuselabs/openraft/1225#-O3XLXXN0Z949VfMIM0w:-O3b2R-79OnmKs69aSgQ:b-isqdch) ([raw file](https://github.com/datafuselabs/openraft/blob/2411bd7a0a7a4df402dc777bdec2cd4cb47a900e/openraft/src/core/heartbeat/worker.rs#L87)):*
<details><summary><i>Previously, drmingdrmer (张炎泼) wrote…</i></summary><blockquote>

Semantically, `prev_log_id` is used to ensure that the log entries in `AppendEntries.entries` are consecutive with the entries on the leader by asserting that `prev_log_id` is equal. Raft ensures that a matching log ID implies all preceding logs match. If `entries` is empty, it is considered consecutive at any index on the follower, and `prev_log_id=None` matches the very first position `None`.

In short AppendEntries with `prev_log_id=None && entries.is_empty()` means to append nothing at the very beginning, which is always valid.

[https://github.com/datafuselabs/openraft/blob/ada27b447a2a89ca97a6ed5061e3cc724883de85/openraft/src/engine/handler/following_handler/mod.rs#L116](https://github.com/datafuselabs/openraft/blob/ada27b447a2a89ca97a6ed5061e3cc724883de85/openraft/src/engine/handler/following_handler/mod.rs#L114-L131)
</blockquote></details>

Strictly speaking, this was previously undefined on the API level. Now it is defined, thanks.

**Comment by SteveLauC** [comment]
Perhaps we should include this change in the [architecture](https://github.com/datafuselabs/openraft/blob/main/openraft/src/docs/internal/architecture.md) doc?

**Comment by drmingdrmer** [comment]
> Perhaps we should include this change in the [architecture](https://github.com/datafuselabs/openraft/blob/main/openraft/src/docs/internal/architecture.md) doc?

Make sense

**Comment by ariesdevil** [comment]
And some notes need to be mentioned in the doc. E.g, the node disk may have been damaged, but the heartbeat has been normal.
